fuzz/fuzz_compile.c

```diff
@@ -0,0 +1,105 @@
+/* Copyright 2020 Google Inc.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+ http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ */
+
+#include "quickjs.h"
+#include "quickjs-libc.h"
+#include "cutils.h"
+
+#include <stdint.h>
+#include <stdio.h>
+
+static int nbinterrupts = 0;
+
+// handle timeouts from infinite loops
+static int interrupt_handler(JSRuntime *rt, void *opaque)
+{
+    nbinterrupts++;
+    return (nbinterrupts > 100);
+}
+
+int LLVMFuzzerTestOneInput(const uint8_t *Data, size_t Size) {
+    if (Size == 0)
+        return 0;
+
+    JSRuntime *rt = JS_NewRuntime();
+    // 64 Mo
+    JS_SetMemoryLimit(rt, 0x4000000);
+    // 64 Kb
+    JS_SetMaxStackSize(rt, 0x10000);
+    JSContext *ctx = JS_NewContext(rt);
+    JS_AddIntrinsicBigFloat(ctx);
+    JS_AddIntrinsicBigDecimal(ctx);
+    JS_AddIntrinsicOperators(ctx);
+    JS_EnableBignumExt(ctx, 1);
+    JS_SetModuleLoaderFunc(rt, NULL, js_module_loader, NULL);
+    JS_SetInterruptHandler(JS_GetRuntime(ctx), interrupt_handler, NULL);
+    js_std_add_helpers(ctx, 0, NULL);
+
+    uint8_t *NullTerminatedData = (uint8_t *)malloc(Size + 1);
+    memcpy(NullTerminatedData, Data, Size);
+    NullTerminatedData[Size] = 0;
+    JSValue obj;
+    obj = JS_Eval(ctx, (const char *)NullTerminatedData, Size, "<none>", JS_EVAL_FLAG_COMPILE_ONLY | JS_EVAL_TYPE_GLOBAL | JS_EVAL_TYPE_MODULE);
+    free(NullTerminatedData);
+    //TODO target with JS_ParseJSON
+    if (JS_IsException(obj)) {
+        JS_FreeContext(ctx);
+        JS_FreeRuntime(rt);
+        return 0;
+    }
+    size_t bytecode_size;
+    uint8_t* bytecode = JS_WriteObject(ctx, &bytecode_size, obj, JS_WRITE_OBJ_BYTECODE);
+    JS_FreeValue(ctx, obj);
+    if ( !bytecode ) {
+        JS_FreeContext(ctx);
+        JS_FreeRuntime(rt);
+        return 0;
```

Comment:
(here and elsewhere) Shoudln't this return != 0 on error?

Reply:
It's a tricky question, since from fuzzing perspective every test is useful that increases the coverage, even if it triggers an error recovery code path. On the other hand, if `LLVMFuzzerTestOneInput` returns a non-zero value (usually -1), then it instructs libFuzzer to discard the test case (i.e., do not save it into the corpus, even if it triggers a new code path). There is room for experiment here, but I think the most beneficial approach in this case is to keep everything that discovers new coverage.